src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+        args = {
+            "_reportTimestamp": cdtp.ref_timestamp,
+            "_timeElapsed": cdtp.seconds_elapsed_since_last_report,
+            "_clValidators": len(cdtp.ref_lido_validators),
+            "_clBalance": self.w3.to_wei(ref_lido_balance, 'gwei'),
+            "_withdrawalVaultBalance": ref_withdrawal_vault_balance,
+            "_elRewardsVaultBalance": 0,
+            "_lastFinalizableRequestId": 0,
+            "_simulatedShareRate": 0,
+        }
+        before_report_total_pooled_ether = self.w3.lido_contracts.lido.functions.totalSupply().call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        after_report_total_pooled_ether, _ = self.w3.lido_contracts.lido.functions.handleOracleReport(**args).call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        logger.info({
+            "msg": "Simulate 'handleOracleReport' contract function call",
+            "call_args": args,
+            "before_call": before_report_total_pooled_ether,
+            "after_call": after_report_total_pooled_ether,
+        })
+
+        frame_cl_rebase = self.w3.from_wei(after_report_total_pooled_ether - before_report_total_pooled_ether, 'gwei')
+        logger.info({"msg": f"Simulated CL rebase for frame: {frame_cl_rebase} Gwei"})
+
+        return frame_cl_rebase
+
+    def _is_high_midterm_slashing_penalty(self, cdtp: CommonDataToProcess, cl_rebase: Gwei) -> bool:
+        logger.info({"msg": "Detecting high midterm slashing penalty"})
+        all_slashed_validators = self._not_withdrawn_slashed_validators(cdtp.ref_all_validators, cdtp.ref_epoch)
+        lido_slashed_validators: list[Validator] = [
+            v.validator for v in self.w3.lido_validators.filter_lido_validators(cdtp.lido_keys, all_slashed_validators)
+        ]
+        logger.info({"msg": f"Slashed: {len(all_slashed_validators)=} | {len(lido_slashed_validators)=}"})
+
+        # If no one lido in current slashed validators - no need to bunker
+        if not lido_slashed_validators:
+            return False
+
+        # Fill per epoch buckets by possible slashed epochs
+        per_epoch_buckets = defaultdict(list[Validator])
+        for validator in all_slashed_validators:
+            possible_epochs_with_slashings = self._detect_slashing_epoch_range(validator, cdtp.ref_epoch)
+            for epoch in possible_epochs_with_slashings:
+                per_epoch_buckets[epoch].append(validator)
+
+        min_bucket_epoch = min(per_epoch_buckets.keys())
+
+        # Iterate through per_epoch_buckets and calculate lido midterm penalties for each bucket
+        per_epoch_lido_midterm_penalties = defaultdict(int)
+        # We should calculate total_balance for each bucket, but we do it once for all per_epoch_buckets
+        total_balance = self._calculate_total_effective_balance(cdtp.ref_all_validators)
+        for epoch, slashed_validators in per_epoch_buckets.items():
```

Comment:
Hope we will never got into bunker mode)

Reply:
I'm afraid it could take a lot of time to calculate everything

Reply:
"big problems require big calculations" :smiling_face_with_tear: 
But actually only requests to states will take a lot of time, I guess